### PR TITLE
Add Story runner

### PR DESCRIPTION
**Pull request by eccles**
The ability to run scenarios from a yaml file, json file or
python dictionary. Thei makes it much easier to execute
predefined scenarios.

Signed-off-by: Paul Hewlett <phewlett76@gmail.com>

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -23,63 +23,246 @@ Example
...
+      # note the values to the events.create method are string representations of boolean
+      # and numbers
+      - operation: events.create
+        to_print: Create Event adding 3 rads of radiation to bag 1, increasing its weight by 1kg.
+        asset_name: radiation bag 1
+        args:
+          - operation: Record
+            behaviour: RecordEvidence
+          - arc_description: add waste to bag
+            arc_evidence: see attached report
```
Comment (JAG-UK) [inline README.rst:260]: This only makes sense if there are attachments on the Event. Please either add attachments to this example or alter thee evidence string to something more appropriate.

File: archivist/errors.py [hunk 2]
```diff
@@ -34,8 +34,8 @@ class ArchivistUnwithdrawnError(ArchivistError):
     """Sbom failed to be withdrawn after fixed timeout"""
 
 
-class ArchivistIllegalArgumentError(ArchivistError):
```
Comment (JAG-UK) [inline archivist/errors.py:37]: Why was this removed? Don't we need both?

File: archivist/parser.py [hunk 3]
```diff
@@ -88,6 +94,15 @@ def common_parser(description):
...
     )
+    parser.add_argument(
+        "-n",
+        "--namespace",
+        type=str,
+        dest="namespace",
+        action="store",
+        default=None,
+        help="namespace of item population",
+    )
```
Comment (JAG-UK) [inline archivist/parser.py:105]: This argument already exists in various samples (Wavestone, for example) in the `archivist_samples` repo which already use `common_parser`. Won't this cause errors if we add in both places? Should have an Issue to remove namespace from the individual samples and investigate whether there's any knock-on fixes to be made in the sample code to use the standard namespace

File: archivist/runner.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+    def __call__(self, config: Dict):
+        """
+        The dict config contains a list of `operations` to be performed serially, e.g.
+
+        ```
+        "operations": [
+            {
+                "operation": "assets.create",
+                "to_print": 10,
```
Comment (JAG-UK) [inline archivist/runner.py:45]: Would be more helpful to show an example of a text prompt here, since that's what `to_print` is for.

File: archivist/runner.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                },
+            }
+        ]
+        ```
+
+        where:
+         - `operation` is the operation to perform (name of function)
+         - `to_print` is what to print to the console.
+         - `wait_time` is time to wait before running the operation.
+         - `attrs are the asset's attributes
```
Comment (JAG-UK) [inline archivist/runner.py:64]: closing backtick

File: docs/runner/events.create.rst [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    operations:
+      - operation: events.create
+        to_print: Create Event requesting EV pump 1 needs maintenance.
+        asset_name: ev pump 1
+        args:
+          - operation: Record
+            behaviour: RecordEvidence
+          - arc_correlation_value: EV Maintenance 1
+            arc_description: request maintenance
+            arc_evidence: see attached report
```
Comment (JAG-UK) [inline docs/runner/events.create.rst:24]: Again probably better to have a self-explanatory string here (like 'Service cycles exceeded') if we're not actually adding attachments to the Event

File: functests/test_resources/dynamic_tolerance_story.yaml [hunk 7]
```diff
@@ -1,153 +1,157 @@
...
+        arc_display_name: ev pump 1
         ev_pump: "true"
       confirm: true
 
-
   # create some policies
-  - operation: CREATE_COMPLIANCE_POLICY_DYNAMIC_TOLERANCE
+  - operation: compliance_policies.create_from_data
     to_print: Create a compliance policy that checks an EV pump maintenance requests are serviced within a reasonable time frame.
-    policy_id: ev maintenance policy 1
```
Comment (JAG-UK) [inline functests/test_resources/dynamic_tolerance_story.yaml:15]: We've lost the friendly name reference to this policy. Don't we need that? eg for deleting it in clean-up at the end?

**Review by JAG-UK — CHANGES_REQUESTED** [review]
Very lovely. A few fairly minor things to look at

File: archivist/parser.py (hunk 3 above)
Comment (eccles) [reply archivist/parser.py:105]: archivist-samples currently uses a copy of this so no problems enviaged.

File: archivist/errors.py (hunk 2 above)
Comment (eccles) [reply archivist/errors.py:37]: Turned out that IllegalArgument was not used anywhere. I would assume that this may change whe dataclasses are introduced

File: functests/test_resources/dynamic_tolerance_story.yaml (hunk 7 above)
Comment (eccles) [reply functests/test_resources/dynamic_tolerance_story.yaml:15]: I could not find it used anywhere so removed it...

File: README.rst [hunk 8]
```diff
@@ -23,63 +23,246 @@ Example
...
-    }
+            # it does not start with arc_
+        }
+        #
+        # Select the mechanism used to prove evidence for the asset.  If the selected proof
+        # mechanism is not enabled for your tenant then an error will occur.
+        # If unspecified then SIMPLE_HASH is used.
+        # proof_mechanism = ProofMechanism.KHIPU.name
+        props = {
+            "proof_mechanism": ProofMechanism.KHIPU.name,
```
[the remark this answers is not in the document]
Comment (eccles) [reply README.rst]: fixed

File: README.rst (hunk 1 above)
Comment (eccles) [reply README.rst]: I have used the example in docs.rkvst.com for standard RecordEvidence event.

File: archivist/runner.py (hunk 4 above)
Comment (eccles) [reply archivist/runner.py]: fixed

File: archivist/runner.py (hunk 5 above)
Comment (eccles) [reply archivist/runner.py]: fixed

**Review by JAG-UK — APPROVED** [review]
